src/net_processing.cpp

```diff
@@ -969,10 +957,10 @@ void EraseOrphansFor(NodeId peer)
 {
     LOCK(g_cs_orphans);
     int nErased = 0;
-    std::map<uint256, COrphanTx>::iterator iter = mapOrphanTransactions.begin();
+    auto iter = mapOrphanTransactions.begin();
     while (iter != mapOrphanTransactions.end())
     {
-        std::map<uint256, COrphanTx>::iterator maybeErase = iter++; // increment to avoid iterator becoming invalid
+        const auto maybeErase = iter++;
         if (maybeErase->second.fromPeer == peer)
         {
             nErased += EraseOrphanTx(maybeErase->second.tx->GetHash());
```

Comment:
There is a subtle difference between deleting elements from `std::map` and `std::unordered_map` while iterating on the container - with `std::unordered_map` the order of the elements may change. This is fixed in C++14: https://en.cppreference.com/w/cpp/container/unordered_map/erase
> The  order of the elements that are not erased is preserved. (This makes it  possible to erase individual elements while iterating through the  container.) (since C++14)

For example, in C++11 the following may happen:
* the `std::unordered_map` contains `A`, `B`, `C`, `D`
* we have an iterator pointing to `B`, we copy it, advance it to `C` and erase `B` via the saved copy
* now the advanced operator is not invalidated and still points to `C`
* however the order of the elements in the container is now `D`, `C`, `A`, so advancing the iterator until `end()` will work, but overall we would have visited `A` twice and not visited `D`.

A possible workaround is to pile the iterators in a `std::vector` and only delete them after finishing of the iteration over the `std::unordered_map`.

Reply:
@vasild and me have discussed this concern on IRC. I think as _"Other iterators ... are not invalidated."_ it does not matter whether the internal element order is changed while iterating over `std::unordered_map`.

Maybe our C++ connoisseurs could add their opinions here?

Reply:
If you have an valid iterator without order being preserved, I think you could have:

1. implicit ordering = A, B, C, D, E ; iterator X points at A
2. X++; X points at B
3. X++; X points at C
4. delete B
5. new implicit ordering = E A C B ; iterator X still points at C
6. X++; now points at B, which was already seen
7. X++; now points at end() and we never saw E

https://stackoverflow.com/questions/38468844/erasing-elements-from-unordered-map-in-a-loop/38469800 suggests that all C++11 compilers probably already did the C++14 behaviour (I imagine it's hard to do efficient iterators that remain valid on erasure when you're also reordering elements) so probably hard to find a counterexample, though.

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/19374#issuecomment-651012944).